src/main/kotlin/exercise4/task2/TransactionalBankAccount.kt

```diff
@@ -0,0 +1,225 @@
+
+    fun getAllFailedDeposits() = getAllFailedTransactions().filter { it.getTransactionType() == TransactionType.DEPOSIT }
+    fun getAllFailedWithdrawals() = getAllFailedTransactions().filter { it.getTransactionType() == TransactionType.WITHDRAW }
+    fun getAllSuccessfulDeposits() = getAllSuccessfulTransactions().filter { it.getTransactionType() == TransactionType.DEPOSIT }
+    fun getAllSuccessfulWithdrawals() = getAllSuccessfulTransactions().filter { it.getTransactionType() == TransactionType.WITHDRAW }
+
+    override fun deposit(amount: Double): Boolean {
+        val status = if (super.deposit(amount)) TransactionStatus.SUCCESS else TransactionStatus.FAILURE
+        transactions.add(
+            Transaction(
+                TransactionType.DEPOSIT,
+                amount,
+                balance - amount,
+                balance,
+                status
+            )
+        )
+        return status == TransactionStatus.SUCCESS
+    }
+
+    override fun withdraw(amount: Double): Boolean {
+        val transactionStatus: TransactionStatus = if (super.withdraw(amount)) TransactionStatus.SUCCESS
+        else TransactionStatus.FAILURE
+        val newTransaction = Transaction(TransactionType.WITHDRAW,
+                            amount,
+                            balance + amount,
+                            balance,
+                            transactionStatus)
+        transactions.add(newTransaction)
+        return transactionStatus == TransactionStatus.SUCCESS
+    }
+
+    override fun displayAccountInfo() {
+        println(
+            """
+                Account Holder: $accountHolderName
+                Account Number: $accountNumber
+                Balance: $balance
+                
+                Transactions:
+            """.trimIndent())
+        if (transactions.isEmpty()) println("No transactions recorded.\n".indent(4))
+        else transactions.forEach {
+                println(
+                    """        
+                      Transaction Date: ${it.getTransactionDate().prettyPrint()}
+                      Transaction Type: ${it.getTransactionType()}
+                      Amount: ${it.getAmount()}
+                      Old Balance: ${it.getOldBalance()}
+                      New Balance: ${it.getNewBalance()}
+                      Status: ${it.getStatus()}\n
+                    """.trimIndent().indent(4))
+            }
+    }
+}
+
+class Transaction (
+    private val transactionType: TransactionType,
+    private val amount: Double,
+    private val oldBalance: Double,
+    private val newBalance: Double,
+    private val transactionStatus: TransactionStatus,
+    private val transactionDate: LocalDateTime = currentTime
+) {
+    fun getTransactionType() = transactionType
+    fun getAmount() = amount
+    fun getOldBalance() = oldBalance
+    fun getNewBalance() = newBalance
+    fun getStatus() = transactionStatus
+    fun getTransactionDate() = transactionDate
+}
```

Comment:
This should become a simple data class with public read-only properties:
```kotlin
data class Transaction (
    val transactionType: TransactionType,
    val amount: Double,
    val oldBalance: Double,
    val newBalance: Double,
    val transactionStatus: TransactionStatus,
    val transactionDate: LocalDateTime = currentTime
)
```
There is no need to specify getters, because getters are encapsulated in properties: 

`var` property == getter + setter + field
`val` property == getter + field